app/org/sagebionetworks/bridge/services/ExternalIdService.java

```diff
@@ -69,22 +71,36 @@ public void addExternalIds(Study study, List<String> externalIdentifiers) {
         externalIdDao.addExternalIds(study.getStudyIdentifier(), externalIdentifiers);
     }
     
-    public void reserveExternalId(Study study, String externalIdentifier) {
+    public void reserveExternalId(Study study, String externalIdentifier, String healthCode) {
         checkNotNull(study);
-        checkArgument(isNotBlank(externalIdentifier));
+        checkNotNull(healthCode);
+        
+        if (study.isExternalIdValidationEnabled() && externalIdentifier != null) {
+            ParticipantOptionsLookup lookup = optionsService.getOptions(healthCode);
+            String existingId = lookup.getString(EXTERNAL_IDENTIFIER);
 
-        if (study.isExternalIdValidationEnabled()) {
-            externalIdDao.reserveExternalId(study.getStudyIdentifier(), externalIdentifier);
+            if (!externalIdentifier.equals(existingId)) {
+                externalIdDao.reserveExternalId(study.getStudyIdentifier(), externalIdentifier);    
+            }
         }
     }
     
     public void assignExternalId(Study study, String externalIdentifier, String healthCode) {
         checkNotNull(study);
-        checkArgument(isNotBlank(externalIdentifier));
-        checkArgument(isNotBlank(healthCode));
+        checkNotNull(healthCode);
         
         if (study.isExternalIdValidationEnabled()) {
-            externalIdDao.assignExternalId(study.getStudyIdentifier(), externalIdentifier, healthCode);
+            ParticipantOptionsLookup lookup = optionsService.getOptions(healthCode);
+            String existingExternalId = lookup.getString(EXTERNAL_IDENTIFIER);
+
+            if (isBlank(existingExternalId)) {
+                if (isNotBlank(externalIdentifier)) {
+                    externalIdDao.assignExternalId(study.getStudyIdentifier(), externalIdentifier, healthCode);
+                }
+            } else if (!existingExternalId.equals(externalIdentifier)) {
+                throw new BadRequestException(
+                        "External ID cannot be changed or removed after assignment.");
+            }
```

Comment:
This logic is duplicated between reserve() and assign(). If we change the semantics so you always have to call reserve() before assign(), we could simplify this logic. This means, by the time we get to assign(), we know that we aren't trying to change or remove an existing external ID.